### PR TITLE
Machine Learning Experimentation

### PR DESCRIPTION
**Pull request by carlosmspk**
Experimentation code based on Machine Learning theory acquired from https://www.youtube.com/watch?v=NWONeJKn6kc&t=21210s
Topics:
- Linear and Logistic Regression
- Support Vector Machines
- Principal Component Analysis
- Decision Trees
- Ensemble Learning (incomplete, text only)

### CODE REVIEW

File: DecisionTrees/main.ipynb [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+      "0.9999999999999999\n"
+     ]
+    }
+   ],
+   "source": [
+    "from math import log2\n",
+    "\n",
+    "def entropy(values):\n",
+    "    total = sum(values)\n",
+    "    result = 0\n",
```
Comment (andre-exd) [inline DecisionTrees/main.ipynb:52]: Can you make result be the first variable on this function?

File: DecisionTrees/main.ipynb [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+    "    for val in values:\n",
+    "        if val == 0:\n",
+    "            continue\n",
+    "        prob = val/total\n",
+    "        result += prob*log2(prob)\n",
+    "    result *= -1\n",
+    "    return max(0.0,result)\n",
```
Comment (andre-exd) [inline DecisionTrees/main.ipynb]: I guess that result variable is being abused. Ideally result would be assigned once and returned at the end of a function. Please make sure last line of this function is return result

File: DecisionTrees/main.ipynb [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+    "from math import log2\n",
+    "\n",
+    "def entropy(values):\n",
+    "    total = sum(values)\n",
+    "    result = 0\n",
+    "    for val in values:\n",
+    "        if val == 0:\n",
+    "            continue\n",
+    "        prob = val/total\n",
+    "        result += prob*log2(prob)\n",
```
Comment (andre-exd) [inline DecisionTrees/main.ipynb]: Can you a comment specifying what is being done here

File: DecisionTrees/main.ipynb [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+    "    result *= -1\n",
```
Comment (andre-exd) [inline DecisionTrees/main.ipynb]: Can you add a comment here?

File: DecisionTrees/main.ipynb [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+    "\n",
+    "print (entropy([100, 0])) # No entropy\n",
+    "print (entropy([5, 5])) # full entropy\n",
+    "print (entropy([9, 5])) # high entropy, but there's a tendency\n",
+    "print (entropy([5, 5, 5])) # full entropy for 3 classes\n",
+    "\n",
+    "#if we want to normalize entropy:\n",
+    "def normalized_entropy(values):\n",
+    "    total = sum(values)\n",
+    "    result = 0\n",
```
Comment (andre-exd) [inline DecisionTrees/main.ipynb:75]: Move result to top

File: DecisionTrees/main.ipynb [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+    "    for val in values:\n",
+    "        if val == 0:\n",
+    "            continue\n",
+    "        prob = val/total\n",
+    "        result += prob*log2(prob)\n",
```
Comment (andre-exd) [inline DecisionTrees/main.ipynb]: Add comments

File: Regression/main.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+from sklearn.datasets import load_boston
+from sklearn.model_selection import train_test_split
+from sklearn.linear_model import LinearRegression
+from sklearn.metrics import mean_squared_error
+from scipy import stats
+
+show_plots = False
+
+load_boston = load_boston()
+X, y = load_boston.data, load_boston.target
```
Comment (andre-exd) [inline Regression/main.py]: Shouldn't X be x instead?

File: Regression/main.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+data = pd.DataFrame(X, columns=load_boston.feature_names)
+data["SalePrice"] = y
+print(data.shape)
+print(data.info())
+
+if show_plots:
+    # We can inspect every single variable's relationship to all others. But in large amounts of features, that's hardly practical
+    # sns.pairplot(data, height=2.5)
```
Comment (andre-exd) [inline Regression/main.py]: Can you removed this unused code?

File: Regression/main.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+if show_plots:
+    sns.displot(data["SalePrice"])
+    plt.show()
+    stats.probplot(data["SalePrice"], plot=plt)
+    plt.show()
+    sns.heatmap(corr, annot=True, cmap = plt.cm.seismic)
+    plt.show()
+
+# Train test split
+X = data.drop("SalePrice", axis=1)
```
Comment (andre-exd) [inline Regression/main.py]: x?

File: Regression/main.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+y = data["SalePrice"]
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
```
Comment (andre-exd) [inline Regression/main.py]: x_train and x_test?

File: Regression/main.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+print(f"{mse = }")
+
+cor_target = abs(corr["SalePrice"])
+relevant_features = cor_target[cor_target > .2]
+names = [index for index, _ in relevant_features.iteritems()]
+
+names.remove("SalePrice")
+print(data.shape[1]-1 - len(names), "removed feature")
+
+# we can then use the "new" dataset with the removed feature and we get one less "noisy" variable that is introducing meaningless data
```
Comment (andre-exd) [inline Regression/main.py]: What is usage of this comment? Seems a lost comment

File: SVM/linSVM [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+distance from hyperplane to margin is given as 2/|W|
+
+Thus, the goal is to maximize margin (minimize |W|) subject to yi * (W^T*X - b) >= 1
+This equation only makes sense if yi is either 1 or -1 (not 0). if yi could be 0, all 0 points would obey the restriction above.
+
+Soft Margin SVM:
+with soft margin we allow points to be misclassified, and simply incorporate misclassified points in the cost function
+
+cost function:
+1/n + sum(max(0, 1-yi(W^T*X - b))) + λ|W|^2
```
Comment (andre-exd) [inline SVM/linSVM]: Can you add a new line to the end of this file?

File: SVM/main.ipynb [hunk 13]
```diff
@@ -0,0 +1,437 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [
+    {
+     "ename": "ModuleNotFoundError",
```
Comment (andre-exd) [inline SVM/main.ipynb]: Should this errors be presented on this notebook?

File: SVM/main.ipynb [hunk 14]
```diff
@@ -0,0 +1,437 @@
...
+      "(2625, 4)\n",
+      "(657, 4)\n",
+      "(2625,)\n",
+      "(657,)\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Problem example, from all data, predict \"Close\" (whether stocks will go up or down)\n",
+    "X = data.drop(\"Close\", axis=1)\n",
```
Comment (andre-exd) [inline SVM/main.ipynb:133]: x?

File: SVM/main.ipynb [hunk 15]
```diff
@@ -0,0 +1,437 @@
...
+    "y = data[\"Close\"]\n",
+    "\n",
+    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)\n",
```
Comment (andre-exd) [inline SVM/main.ipynb:136]: x_train?

File: SVM/nonlinSVM [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+k(x,z) = 
+exp(- ||x-z||^2 / (2^(sigma^2)))
+
+> Primal problem:
+for each i in (1,2,...,n)
+Ei = max(0, 1.yi(W^T*X - b))
+
+minimize 1/n sum( Ei + λ|W|^2)
+subject to: yi(W^T*X - b) >= 1 - Ei
+Ei >= b, for all i
```
Comment (andre-exd) [inline SVM/nonlinSVM]: Add new line to EOF

**Review by andre-exd — CHANGES_REQUESTED** [review]

File: DecisionTrees/main.ipynb (hunk 3 above)
Comment (andre-exd) [reply DecisionTrees/main.ipynb]: Seems like you missed this comment

**Review by andre-exd — CHANGES_REQUESTED** [review]

File: DecisionTrees/main.ipynb (hunk 3 above)
Comment (carlosmspk) [reply DecisionTrees/main.ipynb]: Are you referring to the last line of code? Something like this:
```suggestion
    "        # sum this value's entropy contribution to overall entropy",
    "        result += prob*log2(prob)\n",
```

Comment (andre-exd) [reply DecisionTrees/main.ipynb]: Yes. Can you also apply to other lines of result and leave last line only with return result?

**Review by andre-exd — APPROVED** [review]

**Pull request merged**